### ISSUE TITLE
Lack of speed up on OSX Sierra (10.12.6)

### ISSUE THREAD

**Issue by mrava87** [pyFFTW/pyFFTW#258]
Hello,
I am trying to use pyFFTW to speed up some of the routines containing fft in this package (https://pylops.readthedocs.io). So far I seem to be able to install ``pyFFTW`` in three different ways:
* pip 
  ```
  pip install pyfftw
  ```
* pip 
  ``` 
  conda install -c conda-forge pyfftw
  ```
* local
   ``` 
   python setup.py build_ext --inplace
   python setup.py develop
   ```
and I always seem to succeed in the installation process but then pyFFTW seems to be slower than numpy.

I have tried to kept track of all installations steps here https://github.com/mrava87/pyFFTW_playground/blob/master/README.md and https://github.com/mrava87/pyFFTW_playground/blob/master/LOCALINSTALL.md, and the environment I use is created from https://github.com/mrava87/pyFFTW_playground/blob/master/environment.yml file, hopefully someone can see where I am doing something wrong. Here is some examples (mostly taken from your documentation) where I don't see any speed up :( https://github.com/mrava87/pyFFTW_playground/blob/master/examples.ipynb

Note that your README.md file is quite confusing when it comes to 
```
export DYLD_LIBRARY_PATH=/usr/local/lib
export LDFLAGS="-L/usr/local/lib"
export CFLAGS="-I/usr/local/include"
```
is this required only before ``pip`` or also before ``conda`` and local installation? I always add those to my environment before running any of the three, maybe I am doing something wrong here?

Thank you!

**Comment by grlee77** [comment]
A few possible things to look into:

1.) How did you install NumPy? If you installed NumPy from conda's default channel it is probably using the Intel MKL package `mklfft` which is much faster than the default `numpy.fft.fftpack` implementation. If you do have MKL fft and want to compare to fftpack you can use the fftpack versions by importing functions such as `fftn` from `numpy.fft.fftpack` (note that `fftpack` will be replaced by `pocketfft` starting from NumPy 1.17, though)

2.) For pyFFTW, the interfaces and planners default to single threaded operation.  Try:
```Python
import multiprocessing
import pyfftw.config
pyfftw.config.NUM_THREADS = multiprocessing.cpu_count()
```

3.) The pyFFTW numpy interfaces are convenient, but for maximum performance you should use the [builders](https://hgomersall.github.io/pyFFTW/pyfftw/builders/builders.html) to preplan the FFTs (assuming you are going to be doing multiple transforms of the same size/shape). Also try changing the planner effort used to `'FFTW_MEASURE'` and see if that helps.

**Comment by grlee77** [comment]
Regarding your other point about `DYLD_LIBRARY_PATH`, etc, those are used when building the library from the source and help the compiler find the FFTW C libraries and headers.

If you are installing via pip wheels or conda packages, the library has already been built and you should not need to set any of these variables.

**Comment by mrava87** [comment]
> Regarding your other point about `DYLD_LIBRARY_PATH`, etc, those are used when building the library from the source and help the compiler find the FFTW C libraries and headers.
> 
> If you are installing via pip wheels or conda packages, the library has already been built and you should not need to set any of these variables.

In you README file I read:

set temporary environmental variables, such that pyfftw finds fftw:
```
export DYLD_LIBRARY_PATH=/usr/local/lib
export LDFLAGS="-L/usr/local/lib"
export CFLAGS="-I/usr/local/include"
```
Now install pyfftw from pip:
```
pip install pyfftw
```

Is this thus incorrect? Does it mean that if I install from `conda` or `pip` I do **NOT** even need to have FFTW installed on my computer?

**Comment by mrava87** [comment]
Thanks a lot!

> A few possible things to look into:
> 
> 1.) How did you install NumPy? If you installed NumPy from conda's default channel it is probably using the Intel MKL package `mklfft` which is much faster than the default `numpy.fft.fftpack` implementation. If you do have MKL fft and want to compare to fftpack you can use the fftpack versions by importing functions such as `fftn` from `numpy.fft.fftpack` (note that `fftpack` will be replaced by `pocketfft` starting from NumPy 1.17, though)
> 

I indeed use latest conda no `numpy` uses MKL backend :) when you say `want to compare to fftpack ` do you mean compare `numpy.fft` to `numpy.fft.fttpack` or `numpy.fft.fftpack` with `pyfftw`.?

> 2.) For pyFFTW, the interfaces and planners default to single threaded operation. Try:
> 
> ```python
> import multiprocessing
> import pyfftw.config
> pyfftw.config.NUM_THREADS = multiprocessing.cpu_count()
> ```

I have those at the top of my notebook https://github.com/mrava87/pyFFTW_playground/blob/master/examples.ipynb forcing NUM_THREADS to 4 but will give it a try again with multiprocessing :)


> 
> 3.) The pyFFTW numpy interfaces are convenient, but for maximum performance you should use the [builders](https://hgomersall.github.io/pyFFTW/pyfftw/builders/builders.html) to preplan the FFTs (assuming you are going to be doing multiple transforms of the same size/shape). Also try changing the planner effort used to `'FFTW_MEASURE'` and see if that helps.

Good point, will modify my notebook trying to make all comparisons using the `builder` and just compare the effective fft computation. My interest is in finding out if and how faster is pyFFTW to numpy.fft (especially now that numpy uses `mkl`) and the planning part is an overhead I am not afraid of as my library [PyLops](https://github.com/Statoil/pylops) will use fft something like 100 times with same planner :)

BTW, do you know if all your claims about speed-up in https://pyfftw.readthedocs.io/en/latest/source/tutorial.html like in the `Monkeypatching` example were done prior to conda using `mkl` or they are still valid?

**Comment by mrava87** [comment]
Following your suggestions I made this script to test performances using directly FFTW (same applies for builder): https://github.com/mrava87/pyFFTW_playground/blob/master/example.py

I can see that if the number of samples is very small numpy MKL seem to be FFTW but when the dimension increases FFTW gets better and better. And as you mentioned, if you use ``numpy.fft.fftpack`` then things get terribly slow, but I guess this is what you would expect?

**Comment by grlee77** [comment]
Yes, `fftpack` should be the slowest (it can be orders of magnitude slower for large prime sizes), but has the least restrictive license terms. `mklfft` is not officially supported by NumPy or used in their PyPI wheels due at least in part to the licensing issues. Anaconda does patch it into their NumPy for the performance benefits unless the user requests the `nomkl` variant.

As far as environment variables, pyFFTW can be built with either pthreads or OpenMP-based threads (or without either).  The `conda-forge` package uses pthreads and so it will not check the OMP_NUM_THREADS environment variable.  However, you can define PYFFTW_NUM_THREADS environment variable which will get used to determine the default number of threads at package import time.  If interested, [here is the initialization logic used](https://github.com/pyFFTW/pyFFTW/blob/5bf1dda837dcef722ee14b349834a701e4d53592/pyfftw/config.py#L54-L65).

**Comment by mrava87** [comment]
Just to follow up on this, I have included pyFFTW as one of the backends for fft in PyLops (https://pylops.readthedocs.io/en/latest/api/generated/pylops.signalprocessing.FFT.html#pylops.signalprocessing.FFT) and also created a small script aimed at generating quick plots of performance (script: https://github.com/mrava87/pyFFTW_playground/blob/master/benchmark.py, plot: https://github.com/mrava87/pyFFTW_playground/blob/master/benchmarks/mac_book_air_1_3_GHz_Intel_Core_i5__np_fft.png)... I hope to get some experience from various hardware and installations (conda vs pip) of which fft is faster in which scenario :) If you are interested to have something like this in your repo to facilitate comparisons between different fft engines, I could clean up the script and make a PR :)
